draft-how-does-mutant-work.markdown

```diff
@@ -0,0 +1,1630 @@
+
+But let's not concern ourselves with the details of the single mutators but first finish off `Mutator.mutate`. So what does this method actually return?
+
+Let's use the console again.
+First keep in mind what the original node looked like:
+
+```Ruby
+s(:def, :say_hello,
+  s(:args,
+    s(:arg, :name)),
+  s(:if,
+    s(:ivar, :@enabled),
+    s(:dstr,
+      s(:begin,
+        s(:ivar, :@phrase)),
+      s(:str, " "),
+      s(:begin,
+        s(:lvar, :name))), nil))
+```
+
+Now let's take samples of what `Mutator.mutate(node)` returns:
+
+```Ruby
+# [18] pry(#<Mutant::Subject::Method::Instance>)> Mutator.mutate(node).to_a.sample
+
+ s(:def, :say_hello,
+  s(:args,
+    s(:arg, :name)),
+  s(:if,
+    s(:ivar, :@enabled),
+    s(:dstr,
+      s(:begin,
+        s(:send, nil, :phrase)),
+      s(:str, " "),
+      s(:begin,
+        s(:lvar, :name))), nil))
+```
+
+Interesting. See the difference to the original node?
+
+The original:
+
+```Ruby
+s(:dstr,
+  s(:begin,
+    s(:ivar, :@phrase)),
+```
+
+The mutation:
+
+```Ruby
+s(:dstr,
+  s(:begin,
+    s(:send, nil, :phrase)),
+```
+
+This means `Mutant` replaced accessing an instance variable
+
+```Ruby
+"#{@phrase} ...."
+```
+
+with a method call:
+
+```Ruby
+"#{phrase} ...."
+```
+
+This:
+
+```
+s(:send, nil, :phrase)
+```
+
+means "Send the message `phrase` to `self`." The "nil" argument denotes that there is no explicit receiver, so `self` is the implicit one.
+
+You can quickly check this out yourself using the `ruby-parse` executable that comes along with the `Parser` gem (so `gem install parser`):
+
+```
+($) ruby-parse -e '"#{@phrase}"'
+
+(dstr
+  (begin
+    (ivar :@phrase)))
+
+($) ruby-parse -e '"#{phrase}"'
+
+(dstr
+  (begin
+    (send nil :phrase)))
+```
+
+Ok, another roll of the dice:
+
+```Ruby
+#[18] pry(#<Mutant::Subject::Method::Instance>)> Mutator.mutate(node).to_a.sample
+
+=> s(:def, :say_hello,
+  s(:args,
+    s(:arg, :name)),
+  s(:if,
+    s(:true),
+    s(:dstr,
+      s(:begin,
+        s(:ivar, :@phrase)),
+      s(:str, " "),
+      s(:begin,
+        s(:lvar, :name))), nil))
+```
+
+There it is! 
+
+```Ruby
+s(:if,
+  s(:true),
+  s(:dstr,
+    s(:begin,
+      s(:ivar, :@phrase))
+```
+
+That's the mutation from our original example. 
+
+```Ruby
+if true
+  "#{@phrase}#{" "}#{name}"
+end
+```
+
+### Mutating single nodes
+
+Alright, time for a recap. So far we have:
+
+* A [configuration](https://github.com/mbj/mutant/blob/master/lib/mutant.rb#L191)
+* An [environment](https://github.com/mbj/mutant/blob/master/lib/mutant/env.rb)
+* A [bootstrapping process](https://github.com/mbj/mutant/blob/master/lib/mutant/env/bootstrap.rb), that ties them together
+* [Subjects](https://github.com/mbj/mutant/blob/master/lib/mutant/subject.rb) that we intend to test
+* [A static mutator](https://github.com/mbj/mutant/blob/master/lib/mutant/mutator.rb) that takes the method nodes within those subjects
```

Comment:
In what way its static? The entry method is `.mutate` but under the hood lots of instances are created. For a recap like this, I'd not worry about implementation details such as the proxying through a static method, and simply say `A mutator`.